crates/storage/provider/src/post_state.rs

```diff
@@ -413,127 +375,27 @@ impl PostState {
-                storage.wiped = false;
-            }
-        }
-    }
-
     /// Write the post state to the database.
-    pub fn write_to_db<'a, TX: DbTxMut<'a> + DbTx<'a>>(
-        mut self,
-        tx: &TX,
-        first_transition_id: TransitionId,
-    ) -> Result<(), DbError> {
-        // Collect and sort changesets by their key to improve write performance
-        let mut changesets = std::mem::take(&mut self.changes);
-        changesets
-            .sort_unstable_by_key(|changeset| (changeset.transition_id(), changeset.address()));
-
-        // Partition changesets into account and storage changes
-        let (account_changes, storage_changes): (Vec<Change>, Vec<Change>) =
-            changesets.into_iter().partition(|changeset| {
-                matches!(
-                    changeset,
-                    Change::AccountChanged { .. } |
-                        Change::AccountCreated { .. } |
-                        Change::AccountDestroyed { .. }
-                )
-            });
-
+    pub fn write_to_db<'a, TX: DbTxMut<'a> + DbTx<'a>>(self, tx: &TX) -> Result<(), DbError> {
         // Write account changes
-        tracing::trace!(target: "provider::post_state", len = account_changes.len(), "Writing account changes");
+        tracing::trace!(target: "provider::post_state", "Writing account changes");
         let mut account_changeset_cursor = tx.cursor_dup_write::<tables::AccountChangeSet>()?;
-        for changeset in account_changes.into_iter() {
-            match changeset {
-                Change::AccountDestroyed { id, address, old } |
-                Change::AccountChanged { id, address, old, .. } => {
-                    let destroyed = matches!(changeset, Change::AccountDestroyed { .. });
-                    let transition_id = first_transition_id + id;
-                    tracing::trace!(target: "provider::post_state", transition_id, ?address, ?old, destroyed, "Account changed");
-                    account_changeset_cursor
-                        .append_dup(transition_id, AccountBeforeTx { address, info: Some(old) })?;
-                }
-                Change::AccountCreated { id, address, .. } => {
-                    let transition_id = first_transition_id + id;
-                    tracing::trace!(target: "provider::post_state", transition_id, ?address, "Account created");
-                    account_changeset_cursor
-                        .append_dup(transition_id, AccountBeforeTx { address, info: None })?;
-                }
-                _ => unreachable!(),
+        for (block_number, account_changes) in self.account_changes.into_iter() {
+            for (address, info) in account_changes.into_iter() {
+                tracing::trace!(target: "provider::post_state", block_number, ?address, old = ?info, "Account changed");
+                account_changeset_cursor
+                    .append_dup(block_number, AccountBeforeTx { address, info })?;
```

Comment:
clean!

if `info` was None the account was new, if Some, it's changed or destroyed